### PR TITLE
[programmable transaction]  Migrate SplitCoin to SplitCoins

### PR DESCRIPTION
**Pull request by tnowacki**
## Description 
    
- Replace SplitCoin with SplitCoins which takes multiple amounts

## Test Plan 

Added additional tests for `SplitCoins` and adjusted existing ones to work with `SplitCoins` instead of `SplitCoin`

---
If your changes are not user-facing and not a breaking change, you can skip the following section. Otherwise, please indicate what changed, and then add to the Release Notes section as highlighted during the release process.

### Type of Change (Check all that apply)

- [X] user-visible impact
- [X] breaking change for a client SDKs
- [X] breaking change for FNs (FN binary must upgrade)
- [X] breaking change for validators or node operators (must upgrade binaries)
- [X] breaking change for on-chain data layout
- [X] necessitate either a data wipe or data migration

### CODE REVIEW

**Comment by tnowacki** [comment]
Stacked on #9420 #9454

File: crates/sui-adapter-transactional-tests/tests/programmable/transfer_objects.move [hunk 1]
```diff
@@ -1,7 +1,7 @@
 // Copyright (c) Mysten Labs, Inc.
 // SPDX-License-Identifier: Apache-2.0
 
-// tessts various invalid vector instantions for types
+// tests various valid operations involving TransferObjects
```
Comment (tnowacki) [inline crates/sui-adapter-transactional-tests/tests/programmable/transfer_objects.move:4]: whoops, thanks!

File: crates/sui-rosetta/src/unit_tests/operations_tests.rs [hunk 2]
```diff
@@ -44,13 +41,3 @@ async fn test_operation_data_parsing() -> Result<(), anyhow::Error> {
 
     Ok(())
 }
-#[tokio::test]
-async fn test_shorter_bytearray_bug() {
```
Comment (tnowacki) [inline crates/sui-rosetta/src/unit_tests/operations_tests.rs:48]: Why was this removed?

File: crates/sui-adapter-transactional-tests/tests/programmable/split_coins_invalid.move [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+
+
+// use incorrect arg type for split
+//# programmable --sender A --inputs object(109) @C
+//> 0: SplitCoins(Input(0), [Input(1)]);
+
+// use incorrect arg type for split coming from a Move function
+//# programmable --sender A --inputs object(109)
+//> 0: test::m1::ret_one_amount();
+//> 1: SplitCoins(Input(0), [Result(0)]);
```
Comment (tnowacki) [inline crates/sui-adapter-transactional-tests/tests/programmable/split_coins_invalid.move:45]: Could we add one where the first element (or two) are `u64` but the last one isn't?
Maybe also worth checking that `vector<u64>` is invalid

File: crates/sui-adapter-transactional-tests/tests/programmable/split_coins.exp [hunk 4]
```diff
@@ -0,0 +1,60 @@
...
+task 1 'publish'. lines 9-29:
+created: object(106)
+written: object(105)
+
+task 2 'run'. lines 30-30:
+created: object(108)
+written: object(100), object(107)
+
+task 3 'view-object'. lines 32-34:
+Error: task 3, lines 32-32. Unbound fake id 109
```
Comment (tnowacki) [inline crates/sui-adapter-transactional-tests/tests/programmable/split_coins.exp:15]: Looks like the view-object got off by 1 or two in these tests

**Review by tnowacki — COMMENTED** [review]
Tests and everything lgtm! But we should probably get some other eyes on this before landing

File: crates/sui-types/src/programmable_transaction_builder.rs [hunk 5]
```diff
@@ -296,23 +296,32 @@ impl ProgrammableTransactionBuilder {
...
         // to minimize the number of transfers that must be performed
         let mut recipient_map = IndexMap::new();
+        let mut amt_args = vec![];
         for (recipient, amount) in recipients.into_iter().zip(amounts) {
+            let cur = amt_args.len();
             recipient_map
                 .entry(recipient)
                 .or_insert_with(Vec::new)
-                .push(amount);
+                .push(cur);
```
Comment (amnn) [inline crates/sui-types/src/programmable_transaction_builder.rs:312]: nit: `enumerate` instead of calculating `cur` + `or_default` instead of `or_insert_with`?

```suggestion
        for (i, (recipient, amount)) in recipients.into_iter().zip(amounts).enumerate() {
            recipient_map
                .entry(recipient)
                .or_default()
                .push(i);
```

File: sdk/typescript/test/e2e/txn-builder.test.ts
Comment (amnn) [inline sdk/typescript/test/e2e/txn-builder.test.ts:1]: We should add some tests exercising the multiple coin behaviour here as well.

**Review by amnn — COMMENTED** [review]
LGTM as well!  Will leave @tnowacki to give the final stamp as there were some good questions about why that test was removed.  Also I remain unfamiliar with the rosetta side of things, so couldn't give that a thorough review.

File: crates/sui-rosetta/src/unit_tests/operations_tests.rs (hunk 2 above)
Comment (awelc) [reply crates/sui-rosetta/src/unit_tests/operations_tests.rs:48]: @patrickkuo has a fix for what this workaround for originally for and suggested we remove this test altogether as no longer necessary.

File: crates/sui-adapter-transactional-tests/tests/programmable/split_coins.exp (hunk 4 above)
Comment (awelc) [reply crates/sui-adapter-transactional-tests/tests/programmable/split_coins.exp:15]: Yeah, removed one account from `init` and did not notice...

**Comment by tnowacki** [comment]
> LGTM as well! Will leave @tnowacki to give the final stamp as there were some good questions about why that test was removed. Also I remain unfamiliar with the rosetta side of things, so couldn't give that a thorough review.

I lack the power to approve since I opened the PR

**Review by awelc — APPROVED** [review]
